nodes/core/storage/50-file.js

```diff
@@ -130,25 +136,66 @@ module.exports = function(RED) {
                             var packet = node.data.shift()
                             node.wstream.write(packet.data, function() {
                                 node.send(packet.msg);
+                                done();
                             });
 
                         } else {
                             // Dynamic filename - write and close the stream
                             var packet = node.data.shift()
                             node.wstream.end(packet.data, function() {
                                 node.send(packet.msg);
+                                delete node.wstream;
+                                delete node.wstreamIno;
+                                done();
                             });
-                            delete node.wstream;
-                            delete node.wstreamIno;
                         }
                     }
                 }
             }
+        }
+
+        function processQ(queue) {
+            var msg = queue[0];
+            processMsg(msg, function() {
+                queue.shift();
+                if (queue.length > 0) {
+                    processQ(queue);
+                }
+            });
+        }
+
+        this.on("input", function(msg) {
+            var msgQueue = node.msgQueue;
+            if (msgQueue.push(msg) > 1) {
+                // pending write exists
+                return;
+            }
+            processQ(msgQueue);
+            if (node.closing) {
+                closeNode();
+            }
         });
-        this.on('close', function() {
+
+        function closeNode() {
             if (node.wstream) { node.wstream.end(); }
             if (node.tout) { clearTimeout(node.tout); }
             node.status({});
+            node.closing = false;
+        }
+
+        this.on('close', function() {
+            if (node.closing) {
+                // already closing
+                return;
+            }
+            node.closing = true;
+            if (node.msgQueue.length > 0) {
+                // close after queue processed
+                return;
```

Comment:
Returning here will tell the runtime the node has finished closing - which isn't true.

You need to make this an async close handler by accepting a `done` parameter:

```
 this.on('close', function(done) {
```

and call `done` once the queue has been cleared.

Reply:
Thank you for pointing this.  Fixed to invoke callbacks if specified.